Add onKey function for custom key commands for patterns

daslyfe opened the pull request:
$: s("bd(5,8)")
  .onKey("Control:j", x => x.segment(16).color("red"))
  .onKey("Control:i", x => x.fast(2).color("blue"))

yaxu commented:
Looks illegal, but not too dangerous :)

Could it support lowercase and short names for modifiers, e.g. `control` and `ctrl`?

I think it is worth exposing that `on` function as well, perhaps as `key`, to support e.g. `key("Control:j")`? so it can be used for other things like switching between patterns with `sew` (although sew still hasn't been ported to strudel yet #31).

Then maybe `onKey` could be renamed as `whenKey` as shorthand for `when(key(), ...)`.. (not sure if the shorthand is worthwhile then, but no strong feelings either way..)

daslyfe commented:
> Nice! Looks illegal, but not too dangerous :)
> 
> Could it support lowercase and short names for modifiers, e.g. `control` and `ctrl`?
> 
> I think it is worth exposing that `on` function as well, perhaps as `key`, to support e.g. `key("Control:j")`? so it can be used for other things like switching between patterns with `sew` (although sew still hasn't been ported to strudel yet #31).
> 
> Then maybe `onKey` could be renamed as `whenKey` as shorthand for `when(key(), ...)`.. (not sure if the shorthand is worthwhile then, but no strong feelings either way..)

good ideas! updated. I added some common aliases and a link to the key name values in the docs, as well as broke out the function and renamed it

yaxu commented:
Great thanks!

Looking a bit closer I thought `isKeyDown` would return a pattern, but now see it returns a single plain value. I'm surprised this works with `when`, but guess it does because the patternification magic in `register` is taking care of doing things inside the `input` pattern so that isKeyDown works on values rather than patterns.

Relying on this magic though means that `when(isKeyDown("Control:b"), ...)` doesn't work, because `isKeyDown` doesn't work with patterns.

How about exporting this instead of `isKeyDown`?

```
export const keyDown = register('keyDown', (pat) => pat.fmap(isKeyDown));
```

(the fmap is needed because register doesn't patternify the final parameter..)

Then `when(keyDown(...` should work, as well as

```js
$: keyDown("Alt:j").pick([s("bd(5,8)"), s("cp(3,8)")]);
```

daslyfe commented:
> ```js
> $: keyDown("Alt:j").pick([s("bd(5,8)"), s("cp(3,8)")])
> ```

works great! I update the pr

daslyfe commented on packages/core/signal.mjs:
```diff
@@ -654,5 +654,21 @@ export const always = register('always', function (func, pat) {
 
 export const whenKey = register('whenKey', function (input, func, pat) {
-  return pat.when(isKeyDown(input), func);
+  return pat.when(keyDown(input), func);
```
done! gonna close this one out
